core/src/main/java/org/elasticsearch/action/index/IndexRequest.java

```diff
@@ -642,6 +635,11 @@ public void process(MetaData metaData, @Nullable MappingMetaData mappingMd, bool
         }
     }
 
+    /* resolve the routing if needed */
+    public void resolveRouting(MetaData metaData) {
```

Comment:
yes, there are consistency issues here (the one you named and more - those methods are to  unify the logic in bulk processing, updates and normal indexing). I decided not to extend the PR to deal with that. We can open a follow up issue.